De-duplication using macros

maxjakob opened the pull request:
There is a lot of duplicated Pig Latin code in `nerd-stats.pig`, `names_and_entities.pig`, `names_and_entities_low_memory.pig` and some more.

This PR introduces macros to de-duplicate, along with a UDF to normalize URIs.

`names_and_entities_low_memory.pig` is deleted for now. [Line 35 or 36](https://github.com/maxjakob/pignlproc/blob/61badf02d08e7c1df12d53081106d43b0ba55038/examples/indexing/names_and_entities.pig#L35) in `names_and_entities.pig` can be commented out/in. We could bring the separate script back if needed.

Also updates maven-assembly-plugin.

Fixes #5.

maxjakob committed:
update maven-assembly-plugin

- use version to 2.4
- build assembly now with command
  mvn package  (assembly:assembly is deprecated)
- adjusted README.md

maxjakob committed:
added UDF to normalize link targets to DBpedia URIs

maxjakob committed:
added macro file that contains functionality of main scripts

will be used by (at least)
- examples/indexing/names_and_entities.pig
- examples/nerd-stats/nerd-stats.pig

maxjakob committed:
scripts import macros

- nerd-stats.pig and names_and_entities.pig both use macros
- deleted examples/indexing/names_and_entities_low_memory.pig

maxjakob commented:
@jodaiber and @chrishokamp, please have a quick look and comment on which other scripts could benefit from this or similar re-arranging. There are a lot of scripts in indexing/ that I don't know what they do and why. Maybe we can also put some more comments on the top of the scripts.

jodaiber commented on examples/macros/nerd_commons.pig:
```diff
@@ -0,0 +1,188 @@
+DEFINE diskIntensiveNgrams(articles, MAX_NGRAM_LENGTH) RETURNS pageNgrams {
+    -- create *all* ngrams in a bag
+
+    DEFINE ngramGenerator pignlproc.helpers.NGramGenerator('$MAX_NGRAM_LENGTH');
```
We should delete NGramGenerator, as it is out of date:
- it does not use our Spotlight tokenizers
- it does not have the latest redirect definitions
- we should not duplicate

The disk-based method should use the same RestrictedNGramGenerator class (it can do both disk-based and memory-based), but with an empty String instead of the surface form lookup file to tell it it should not keep SFs in memory.

```
DEFINE ngramGenerator pignlproc.helpers.RestrictedNGramGenerator('$MAX_NGRAM_LENGTH', '', '$LOCALE');
```

maxjakob replied:
OK, looks like it already does this when you pass an empty string.

But why would it need the redirect definitions?

jodaiber replied:
Wait, I am sorry, I confused it with the Wikipedia loader. There are not redirect definitions in this one.

```
DEFINE ngramGenerator pignlproc.helpers.RestrictedNGramGenerator('$MAX_NGRAM_LENGTH', '', '$LOCALE');
```

Yes, this is the way it was used in names_and_entites_low_memory.pig, which was exactly the same file except for the changed call to the RestrictedNGramGenerator class.

jodaiber commented:
Otherwise looks great. The statistical backend only uses:
- names_and_entities.pig and names_and_entities.pig.params
- token_counts.pig and token_counts.pig.params

I don't know about the other scripts. Maybe we could move them to an "experimental" folder.

maxjakob committed:
deleted NGramGenerator in favor of RestrictedNGramGenerator

RestrictedNGramGenerator can be initiated with '' for the surface forms file,
which results in similar behavior as NGramGenerator, but with Spotlight
tokenizers.

*BUT* the behavior is different: see failing test!

maxjakob committed:
re-factored RestrictedNGramGenerator

- variables to refer to distributed cache files inside and outside of getCacheFiles()
- Bugfix: loading of tokenizer model with correct variable and in exec()
- some formatting

maxjakob commented:
A test fails now:
OpenNLP's `SimpleTokenizer` does separate _"That's"_ into 3 tokens, `LanguageIndependentStringTokenizer` does not.

jodaiber commented:
Hm, I don't know how much sense these tests makes now, given that the tokenizers are more or less exchangeable assuming they are based on model.Tokenizer (as opposed to always being SimpleTokenizer). This is probably rather something that should be tested in spotlight core.

maxjakob commented:
Maybe we should write some tests in the Spotlight core then. This is definitely an issue with texts like

> Controversy plagues Obama's administration.

if we don't have _Obama's_ in the dictionary.

jodaiber commented:
True. The supervised OpenNLP tokenizer will handle that correctly, so in the English version at run-time it should be tokenized as "Obama 's". 
That's the reason we have to ensure we use the same tokenizers at run-time and in the pig scripts for determining c(sf) and c(annotated | sf); the first French model I created was barely useable because of that. Good that you found the bug with the OpenNLP models.

maxjakob committed:
use nerd_commons macros in token_counts.pig

maxjakob committed:
formatting in token_counts.pig

maxjakob committed:
Fix TestNGramGenerator.

Use English OpenNLP that is already in resources/.
Had to add protected constructor to RestrictedNGramGenerator.

maxjakob commented:
Fixed the test in here by using the English tokenizer model that was already in resources/.
